{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Course Project: What's this gene?\n",
    "\n",
    "In this project you will implement a de-novo sequence assembly algorithm. You will be provided with a small sample of fragments (e.g. from an Illumina type machine) for a part of a well-known protein encoding human gene. Your task is to assemble the reads into a sequence then perform an online [BLAST](https://blast.ncbi.nlm.nih.gov/Blast.cgi) search to find out what the gene is.\n",
    "\n",
    "This project will see you use all of the techniques you will learn along the way in the course. During the course you will have the opportunity to work on the project.\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Background\n",
    "\n",
    "_Sequence Assembly_ is the process of reconstructing a sequence from small fragments (usually the output of a sequencing machine) called \"**reads**\". More imformation on [Wikipedia](https://en.wikipedia.org/wiki/Sequence_assembly).\n",
    "\n",
    "We will use a [\"greedy algorithm\"](https://en.wikipedia.org/wiki/Sequence_assembly#Greedy_algorithm) for sequence assembly. This is because it is straightforward to understand and implement but will also give you good enough results to solve the challenge. The steps of the algorithm are:\n",
    "\n",
    "1. Сalculate pairwise alignments of all fragments.\n",
    "1. Choose two fragments with the largest overlap.\n",
    "1. Merge chosen fragments.\n",
    "1. Repeat step 2 and 3 until only one fragment is left or you cannot merge anymore.\n",
    "\n",
    "In the following code cell, we have supplied the fragments you should assemble the sequence from."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fragments = [\n",
    "    \"GAATTAGATAAATTCAAATTAGACTTAGGAAGGAATGTTCCCAATAGTAGACATAAAAGTCTTCGCACAGTGAAAACTAAAATGGATCAAGCAGATGATG\",\n",
    "    \"TTCTTCAGAAGCTCCACCCTATAATTCTGAACCTGCAGAAGAATCTGAACATAAAAACAACAATTACGAACCAAACCTATTTAAAACTCCACAAAGGAAA\",\n",
    "    \"TGCCTATTGGATCCAAAGAGAGGCCAACATTTTTTGAAATTTTTAAGACACGCTGCAACAAAGCAGATTTAGGACCAATAAGTCTTAATTGGTTTGAAGA\",\n",
    "    \"CTCCACAAAGGAAACCATCTTATAATCAGCTGGCTTCAACTCCAATAATATTCAAAGAGCAAGGGCTGACTCTGCCGCTGTACCAATCTCCTGTAAAAGA\",\n",
    "    \"CAACATTTTTTGAAATTTTTAAGACACGCTGCAACAAAGCAGATTTAGGACCAATAAGTCTTAATTGGTTTGAAGAACTTTCTTCAGAAGCTCCACCCTA\",\n",
    "    \"GGCTGACTCTGCCGCTGTACCAATCTCCTGTAAAAGAATTAGATAAATTCAAATTAGACTTAGGAAGGAATGTTCCCAATAGTAGACATAAAAGTCTTCG\",\n",
    "    \"CAAAGAGAGGCCAACATTTTTTGAAATTTTTAAGACACGCTGCAACAAAGCAGATTTAGGACCAATAAGTCTTAATTGGTTTGAAGAACTTTCTTCAGAA\",\n",
    "    \"AGCAAGGGCTGACTCTGCCGCTGTACCAATCTCCTGTAAAAGAATTAGATAAATTCAAATTAGACTTAGGAAGGAATGTTCCCAATAGTAGACATAAAAG\",\n",
    "    \"TCAGCTGGCTTCAACTCCAATAATATTCAAAGAGCAAGGGCTGACTCTGCCGCTGTACCAATCTCCTGTAAAAGAATTAGATAAATTCAAATTAGACTTA\",\n",
    "    \"AATCTCCTGTAAAAGAATTAGATAAATTCAAATTAGACTTAGGAAGGAATGTTCCCAATAGTAGACATAAAAGTCTTCGCACAGTGAAAACTAAAATGGA\",\n",
    "    \"ATGCCTATTGGATCCAAAGAGAGGCCAACATTTTTTGAAATTTTTAAGACACGCTGCAACAAAGCAGATTTAGGACCAATAAGTCTTAATTGGTTTGAAG\",\n",
    "    \"GAATGTTCCCAATAGTAGACATAAAAGTCTTCGCACAGTGAAAACTAAAATGGATCAAGCAGATGATGTTTCCTGTCCACTTCTAAATTCTTGTCTTAGT\"\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Resources\n",
    "\n",
    "The `project` folder contains a complete Python solution to this problem. It also defines several _uitility_ functions that you can use when building your sequence assembly program. Please avoid looking at the solution before trying to solve each step yourself. However, when using a utility function is suggested, it will be imported and used in the template for you."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "## Part 1: Getting started\n",
    "You already have enough Python knowledge to start implementing our sequence assembly program. Let's start off by breaking down the problem into more managable parts (this is a skill you will develop very quickly while programming).\n",
    "\n",
    "Step 1 says we need to compute [pairwise alignments](https://en.wikipedia.org/wiki/Sequence_alignment#Pairwise_alignment). To do this we need 2 things: a way to score alignments, and a way to generate alignments. An easy way to score an alignment is known as the [_edit distance_](https://en.wikipedia.org/wiki/Levenshtein_distance), which is simply the minimum number of changes (character insertions, deletions, or substitutions) that are required to transform one string into another.\n",
    "\n",
    "### Edit distance: an example\n",
    "Imagine I want to find the edit distance between the strings \"kitten\" and \"sitting\":\n",
    "\n",
    "1. **k**itten → **s**itten (substitution of \"s\" for \"k\")\n",
    "2. sitt**e**n → sitt**i**n (substitution of \"i\" for \"e\")\n",
    "3. sittin → sittin**g** (insertion of \"g\" at the end).\n",
    "\n",
    "So the _edit distance_ between \"kitten\" and \"sitting\" is 3.\n",
    "\n",
    "Your first task is to write the beginnings of a function to compute the edit distance. Instead of a whole string, your input will be 2 characters. Please fill in the template provided for you and ensure it passes the test cases..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def edit_distance(queryA, queryB):\n",
    "    if _:\n",
    "        return _\n",
    "    \n",
    "    return _\n",
    "\n",
    "# Do not change the code below, they represent the test cases. \n",
    "assert edit_distance('A', 'T') == 1, \"I expected 1, got: \" + str(edit_distance('A', 'T'))\n",
    "assert edit_distance('G', 'G') == 0, \"I expected 0, got: \" + str(edit_distance('G', 'G'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Don't worry, you will soon extend this to longer sequences but it's a good start for now.\n",
    "\n",
    "Another task we will need to do from step 3 of our algorithm is \"merge\" fragments. Your next task is to write a merging function that takes 2 fragments and merges them end-to-end in order. You can fill in the template below and ensure it passes the test cases..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def merge(fragA, fragB):\n",
    "    return _\n",
    "    \n",
    "assert merge(\"\", \"ATG\") == \"ATG\", \"I expected ATG, got: \" + str(merge('', 'ATG'))\n",
    "assert merge(\"ATG\", \"\") == \"ATG\", \"I expected ATG, got: \" + str(merge('ATG', ''))\n",
    "assert merge(\"ATG\", \"CCT\") == \"ATGCCT\", \"I expected ATGCCT, got: \" + str(merge('ATG', 'CCT'))\n",
    "assert merge(\"A\", \"TG\") == \"ATG\", \"I expected ATG, got: \" + str(merge('A', 'TG'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Well done! You have now completed the beginnings of the sequence assembly program. It doesn't look like much now but it's a great foundation for tomorrow. Congratulate yourself, you've earned it!\n",
    "\n",
    "---\n",
    "\n",
    "## Part 2: Sequence assembly made easy!\n",
    "![How to draw an owl](https://i.kym-cdn.com/photos/images/newsfeed/000/572/078/d6d.jpg)\n",
    "\n",
    "Now that you can iterate over sequences, it's time to complete your _edit distance_ function to work on sequences longer than a single character. We will use a technique called [_dynamic programming_](https://en.wikipedia.org/wiki/Dynamic_programming). _Dynamic programming_ is a technique that allows you to solve big problems by breaking them into smaller and smaller sub-problems that are eventually \"trivial\". For example, the \"trivial problem\" when computing the edit distance is the comparison of 2 characters (that we solved yesterday). Dynamic further involves remembering (or [_memoising_](https://en.wikipedia.org/wiki/Memoization)) partial solutions as you build them up. So how is a solution \"built up\"? Let's have a look at the \"kitten\"/\"sitting\" example from before:\n",
    "\n",
    "Start by setting up a matrix:\n",
    "![Step 1](images/dp_1.jpg)\n",
    "\n",
    "Label the rows and columns:\n",
    "![Step 2](images/dp_2.jpg)\n",
    "\n",
    "Fill in the \"boundary cases\":\n",
    "![Step 3](images/dp_3.jpg)\n",
    "\n",
    "The general case requires you find the `min()` between three cases:\n",
    "![Recurrence relation](https://wikimedia.org/api/rest_v1/media/math/render/svg/10554aecc5e56da9be4657acd75b9a67b5e8b394)  \n",
    "More information regarding this mathematical expression is explained [here](https://en.wikipedia.org/wiki/Levenshtein_distance).\n",
    "\n",
    "If \"S\" is equal to \"K\" the substitution cost is `0`, otherwise `1`. The insertion cost and the deletion cost is always `1`, so:\n",
    "![General case](images/dp_4.jpg)\n",
    "\n",
    "Then systematically fill out the entire matrix:\n",
    "![Systematically fill in the matrix](images/dp_6.jpg)\n",
    "\n",
    "Once the matrix is complete, the bottom-left corner cell will contain the edit distance:\n",
    "![Edit distance](images/dp_8.jpg)\n",
    "\n",
    "I encourage you to try this for yourself with pencil and paper. Once you have convinced yourself that you understand how it works, you can implement it in Python. To begin, the memoisation matrix can be represented using a list of lists:\n",
    "\n",
    "```python\n",
    "memo = [\n",
    "   # -  K  I  T  T  E  N\n",
    "    [0, 0, 0, 0, 0, 0, 0], # -\n",
    "    [0, 0, 0, 0, 0, 0, 0], # S\n",
    "    [0, 0, 0, 0, 0, 0, 0], # I\n",
    "    [0, 0, 0, 0, 0, 0, 0], # T\n",
    "    [0, 0, 0, 0, 0, 0, 0], # T\n",
    "    [0, 0, 0, 0, 0, 0, 0], # I\n",
    "    [0, 0, 0, 0, 0, 0, 0], # N\n",
    "    [0, 0, 0, 0, 0, 0, 0]  # G\n",
    "]\n",
    "```\n",
    "\n",
    "Lets start by working out how we read and write a value from this matrix. Let's begin by writing a function to write the value to a row and column in this matrix, and a corresponding function to get a value from a row and column in this matrix. Please fill in the functions below and ensure it passes the test cases:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def set_value_in_matrix(row, column, matrix, value):\n",
    "    # Set matrix at (row, column) to value\n",
    "    ...\n",
    "    return matrix\n",
    "    \n",
    "def get_value_from_matrix(row, column, matrix):\n",
    "    # Get value in matrix (row, column)\n",
    "    return _\n",
    "    \n",
    "m = [[0, 1, 2], [3, 4, 5], [6, 7, 8]]\n",
    "assert get_value_from_matrix(0, 1, set_value_in_matrix(0, 1, m, 5)) == 5\n",
    "assert get_value_from_matrix(2, 0, set_value_in_matrix(2, 0, m, 81)) == 81\n",
    "assert get_value_from_matrix(2, 2, set_value_in_matrix(2, 2, m, 112)) == 112"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ok great work, we can read and write to our matrix. Now let's create a matrix of the correct size for our input strings. Remember that the dynamic programming matrix is the size of the input strings + 1. Please fix the function that follows and ensure it passes the tests."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_matrix(inputA, inputB):\n",
    "    row = [0] * _\n",
    "    m = []\n",
    "    for _ in range(_):\n",
    "        m.append(row.copy())\n",
    "    \n",
    "    return m\n",
    "\n",
    "assert create_matrix('','') == [[0]], \"Expected: {[[0]]}, got: \" + str(create_matrix('',''))\n",
    "assert create_matrix('dn', 'a') == [[0, 0], [0, 0], [0, 0]], \"Expected: {[[0, 0], [0, 0], [0, 0]]}, got: \" + str(create_matrix('dn', 'a'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now you can initialise the matrix with the boundary cases (a edit distances against empty strings). Remember that the edit distances along the boundary are just the column or row index. For example, this is a matrix with boundary values initialised for the strings \"kitten\" and \"sitting\":\n",
    "\n",
    "```python\n",
    "memo = [\n",
    "   # -  K  I  T  T  E  N\n",
    "    [0, 1, 2, 3, 4, 5, 6], # -\n",
    "    [1, 0, 0, 0, 0, 0, 0], # S\n",
    "    [2, 0, 0, 0, 0, 0, 0], # I\n",
    "    [3, 0, 0, 0, 0, 0, 0], # T\n",
    "    [4, 0, 0, 0, 0, 0, 0], # T\n",
    "    [5, 0, 0, 0, 0, 0, 0], # I\n",
    "    [6, 0, 0, 0, 0, 0, 0], # N\n",
    "    [7, 0, 0, 0, 0, 0, 0]  # G\n",
    "]\n",
    "```\n",
    "\n",
    "Please fix the function that follows and ensure it passes the tests."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def init_matrix(inputA, inputB):\n",
    "    m = create_matrix(inputA, inputB)\n",
    "    rows = _\n",
    "    cols = _\n",
    "    \n",
    "    for _ in range(rows + 1):\n",
    "        ...\n",
    "        \n",
    "    for _ in range(cols + 1):\n",
    "        ...\n",
    "    \n",
    "    return m\n",
    "\n",
    "assert init_matrix('', '') == [[0]], \"Expected: {[[0]]}, got: \" + str(init_matrix('',''))\n",
    "assert init_matrix('dn', 'a') == [[0, 1], [1, 0], [2, 0]], \"Expected: {[[0, 1], [1, 0], [2, 0]]}, got: \" + str(init_matrix('dn', 'a'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now you're finally ready to complete the implementation of your `edit_distance()` function. You can use your `init_matrix()` function. The function template is given below, please complete and fix the function and ensure it passes the test cases."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def edit_distance(queryA, queryB):\n",
    "    m = init_matrix(queryA, queryB)\n",
    "    rows = _\n",
    "    cols = _\n",
    "\n",
    "    for col in range(1, cols + 1):\n",
    "        for row in range(1, rows + 1):\n",
    "            if queryA[row - 1] == queryB[col - 1]:\n",
    "                cost = 0\n",
    "            else:\n",
    "                cost = 1\n",
    "            \n",
    "            m[row][col] = min([m[row-1][col-1] + cost,\n",
    "                               m[row][col - 1] + 1,\n",
    "                               m[row - 1][col] + 1\n",
    "                              ])\n",
    "            \n",
    "    return _ # Return the lower-right corner of the distance matrix\n",
    "\n",
    "assert edit_distance('A', 'T') == 1, \"Expected: 1, got: \" + str(edit_distance('A', 'T'))\n",
    "assert edit_distance('G', 'G') == 0, \"Expected: 0, got: \" + str(edit_distance('G', 'G'))\n",
    "assert edit_distance('kitten', 'sitting') == 3, \"Expected: 3, got: \" + str(edit_distance('kitten', 'sitting'))\n",
    "assert edit_distance('', '') == 0, \"Expected: 0, got: \" + str(edit_distance('', ''))\n",
    "assert edit_distance('ABCD', 'EFGH') == 4, \"Expected: 4, got: \" + str(edit_distance('ABCD', 'EFGH'))\n",
    "assert edit_distance('ABCD', 'ZBCZ') == 2, \"Expected: 2, got: \" + str(edit_distance('ABCD', 'ZBCZ'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Great work! You can now compute the edit distance between arbitrary strings!\n",
    "Now it's time to generate alignments... luckily you're already most of the way there 😊\n",
    "\n",
    "### What is an alignment?\n",
    "An alignment is an assignment of one-to-one correspondence between elements of a sequence. For this project you are working with sequences of nucleotides, and alignments between sequences of nucleotides are assignments of correspondence between _homologous_ nucleotides. An example of what a (pairwise) alignment looks like:\n",
    "\n",
    "<!-- TACCCC---TCC -->\n",
    "<!-- TA--GGTAATGG -->\n",
    "<pre>\n",
    "<table>\n",
    "    <tr>\n",
    "        <td style=\"background-color:green; padding-left: 0; padding-right: 0\">TA</td><td style=\"background-color:red; padding-left: 0; padding-right: 0\">CC</td><td style=\"background-color:green; padding-left: 0; padding-right: 0\">CC<td style=\"padding-left: 0; padding-right: 0\">---</td><td style=\"background-color:green; padding-left: 0; padding-right: 0\">TCC</td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "        <td style=\"background-color:green; padding-left: 0; padding-right: 0\">TA</td><td style=\"padding-left: 0; padding-right: 0\">--</td><td style=\"background-color:green; padding-left: 0; padding-right: 0\">GG</td><td style=\"background-color:yellow; padding-left: 0; padding-right: 0\">TAA</td><td style=\"background-color:green; padding-left: 0; padding-right: 0\">TGG</td>\n",
    "</table>\n",
    "</pre>\n",
    "\n",
    "In this example, I have highlighted \"assigned correspondences\" in <span style=\"background-color:green\">green</span>, insertions relative to the bottom sequence in <span style=\"background-color:yellow\">yellow</span>, and deletions relative to the bottom sequence in <span style=\"background-color:red\">red</span>. The `-` character is used to denote _no correspondence_.\n",
    "\n",
    "### Generating alignments\n",
    "\n",
    "When you were computing the edit distance above, for each cell, you were selecting a \"direction\" to compute the edit distance from (up, left, or diagonal). It turns out that remembering this direction for each cell is all of the extra information you need to generate the optimal alignment for the _edit distance_ score. Then, generating the alignment is simply a matter of [_backtracking_](https://en.wikipedia.org/wiki/Needleman%E2%80%93Wunsch_algorithm#Tracing_arrows_back_to_origin) from the bottom-right of the matrix to the top-left:\n",
    "\n",
    "![backtracking example](images/backtrack.jpg)\n",
    "\n",
    "You should try this now using pencil and paper. Create a new edit distance matrix for the strings, \"saturday\" and \"sunday\". When you fill in the matrix, draw an arrow marking the direction of the minimum value you computed for each cell. Once you finish, compare your answer with mine by clicking on the arrow below. Note that there is not a unique solution so your arrows might be slightly different to mine, but our edit distances should be the same.\n",
    "\n",
    "<details>\n",
    "    <summary>Solution</summary>\n",
    "    <img src=\"images/backtrack_example.jpg\" />\n",
    "</details>\n",
    "\n",
    "Now let us see how an _alignment_ is generated using this \"direction\" information. Start in the bottom-right cell. If the arrow is `diagonal`, write down the column and row label for that cell. If the arrow is `left` write down the column label and a `-` (gap) character. Finally, if the arrow is `up` write down  the row label and a `-` (gap) character. Then move on the the cell pointed to by the arrow. Repeat this procedure until you arrive at the top-left cell. For example:\n",
    "\n",
    "<table>\n",
    "    <tr>\n",
    "        <td>\n",
    "            <img src=\"images/backtrack_example_1.jpg\"/>\n",
    "        </td>\n",
    "        <td>\n",
    "            <pre>\n",
    "Y\n",
    "Y\n",
    "            </pre>\n",
    "        </td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "        <td>\n",
    "            <img src=\"images/backtrack_example_2.jpg\"/>\n",
    "        </td>\n",
    "        <td>\n",
    "            <pre>\n",
    "AY\n",
    "AY\n",
    "            </pre>\n",
    "        </td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "        <td>\n",
    "            <img src=\"images/backtrack_example_3.jpg\"/>\n",
    "        </td>\n",
    "        <td>\n",
    "            <pre>\n",
    "DAY\n",
    "DAY\n",
    "            </pre>\n",
    "        </td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "        <td>\n",
    "            <img src=\"images/backtrack_example_4.jpg\"/>\n",
    "        </td>\n",
    "        <td>\n",
    "            <pre>\n",
    "NDAY\n",
    "RDAY\n",
    "            </pre>\n",
    "        </td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "        <td>\n",
    "            <img src=\"images/backtrack_example_5.jpg\"/>\n",
    "        </td>\n",
    "        <td>\n",
    "            <pre>\n",
    "UNDAY\n",
    "URDAY\n",
    "            </pre>\n",
    "        </td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "        <td>\n",
    "            <img src=\"images/backtrack_example_6.jpg\"/>\n",
    "        </td>\n",
    "        <td>\n",
    "            <pre>\n",
    "-UNDAY\n",
    "TURDAY\n",
    "            </pre>\n",
    "        </td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "        <td>\n",
    "            <img src=\"images/backtrack_example_7.jpg\"/>\n",
    "        </td>\n",
    "        <td>\n",
    "            <pre>\n",
    "--UNDAY\n",
    "ATURDAY\n",
    "            </pre>\n",
    "        </td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "        <td>\n",
    "            <img src=\"images/backtrack_example_8.jpg\"/>\n",
    "        </td>\n",
    "        <td>\n",
    "            <pre>\n",
    "S--UNDAY\n",
    "SATURDAY\n",
    "            </pre>\n",
    "        </td>\n",
    "    </tr>\n",
    "</table>\n",
    "\n",
    "You should try this for yourself with pencil and paper with the strings \"kitten\" and \"sitting\".\n",
    "\n",
    "If you wish to implement this for yourself, please do as it's a great exercise to practice your Python. If you would prefer to progress without implementing this function you may use the provided `align()` function. Use of this function is demonstrated in the following code cell. Please feel free to look at the implementation and check that you understand it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from solutions.sa import align, edit_distance\n",
    "\n",
    "assert align(\"sunday\", \"saturday\", edit_distance) == (\"s--unday\", \"saturday\"), \"Expected: (\\\"s--unday\\\", \\\"saturday\\\"), got: \" + str(align(\"sunday\", \"saturday\", edit_distance))\n",
    "assert align(\"kitten\", \"sitting\", edit_distance) == (\"kitten-\", \"sitting\"), \"Expected: (\\\"kitten-\\\", \\\"sitting\\\"), got: \" + str(align(\"kitten\", \"sitting\", edit_distance))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Scoring alignments\n",
    "\n",
    "At this point you're getting very close to having all of the required building blocks to build your sequence assembly program. The final building block you require is the ability to merge fragments into longer \"contigs\". But we also need to be able to pick from a set of alignments, which to merge. For this we need to be able to rank alignments.\n",
    "\n",
    "For the task of sequence assembly, we expect to find _overlapping_ fragments. What does this mean? Here are some (idealised) examples of overlapping fragments:\n",
    "\n",
    "<pre>\n",
    "Full sequence: AAGAGGATACTATACCCTAATGAATGTTTGGAACGATTTTTCCTTCAGTGCTCTTTCAGCTCACACTCCATGGCGCATGTTGGTGACG\n",
    "Fragment 1:        GGATACTATACCCTAATGAA\n",
    "Fragment 2:    AAGAGGATACTATACCCTAA\n",
    "Fragment 3:                                                                        CATGGCGCATGTTGGTGACG\n",
    "Fragment 4:                                             CCTTCAGTGCTCTTTCAGCT\n",
    "Fragment 5:                           ATGTTTGGAACGATTTTTCC\n",
    "</pre>\n",
    "\n",
    "An alignment of `Fragment 1` and `Fragment 2` might look like this:\n",
    "<pre>\n",
    "----GGATACTATACCCTAATGAA\n",
    "AAGAGGATACTATACCC---T-AA\n",
    "</pre>\n",
    "\n",
    "This seems like an excellent candidate to merge, but before we do that, let's look at some others. An alignment of `Fragment 1` and `Fragment 5` would look like this:\n",
    "<pre>\n",
    "-GGATACTATACCCTAATGAA\n",
    "ATGTTTGGA-ACGATTTTTCC\n",
    "</pre>\n",
    "\n",
    "Not so great! What about `Fragment 2` and `Fragment 3`:\n",
    "<pre>\n",
    "-A-AGAGGATACTATACCCTAA\n",
    "CATGGCGCAT-GT-TGGTGACG\n",
    "</pre>\n",
    "\n",
    "Also not great! So how can we distinguish between a useful alignment (one that will extend 2 fragments into a contig) and an alignment that is not useful?\n",
    "\n",
    "For our purposes, _sequence identity_ will serve nicely (the supplied project code supplies a `score()` function with some slight modifications to more aggressively select useful alignments). If you wish, you may continue into the next section and use the supplied `score()` function. Otherwise, if you wish to practise here is a template for you to compute sequence identity from an alignment. Ensure your function passes the tests."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sequence_identity(top, bottom):\n",
    "    identity = _\n",
    "    for _ in zip(top, bottom):\n",
    "        ...\n",
    "        \n",
    "    return identity\n",
    "\n",
    "assert sequence_identity(\"\", \"\") == 0, \"Expected: 0, got: \" + str(sequence_identity('', ''))\n",
    "assert sequence_identity(\"A\", \"A\") == 1, \"Expected: 1, got: \" + str(sequence_identity('A', 'A'))\n",
    "assert sequence_identity(\"A\", \"T\") == 0, \"Expected: 0, got: \" + str(sequence_identity('A', 'T'))\n",
    "assert sequence_identity(\"ATGC\", \"CGTA\") == 0, \"Expected: 0, got: \" + str(sequence_identity('ATGC', 'CGTA'))\n",
    "assert sequence_identity(\"ATGC\", \"ATGC\") == 4, \"Expected: 4, got: \" + str(sequence_identity('ATGC', 'ATGC'))\n",
    "assert sequence_identity(\"-AA-TG-C\", \"AA-TTGGG\") == 3, \"Expected: 3, got: \" + str(sequence_identity('-AA-TG-C', 'AA-TTGGG'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merging alignments\n",
    "\n",
    "This is the final building block you require to implement your sequence assembly program. Let's have a look at an ideal example:\n",
    "\n",
    "`----GGATACTATACCCTAATGAA`<br/>\n",
    "`AAGAGGATACTATACCCTAA----`\n",
    "\n",
    "Excepting the gaps at the beginning and the end of the alignment, all of the nucleotides are identical. So a `merge()` for this alignment is:\n",
    "\n",
    "`AAGAGGATACTATACCCTAATGAA`\n",
    "\n",
    "Of course, it's not guaranteed that all corresponding nucleotides are identical so you will have to decide which one to accept (either top or bottom). Complete the `merge()` function below ensuring you pass the tests."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def merge(top, bottom):\n",
    "    ...\n",
    "    \n",
    "assert merge(\"A\", \"A\") == \"A\", \"Expected: 'A', got: \" + str(merge('A', 'A'))\n",
    "assert merge(\"A\", \"T\") == \"A\" or merge(\"A\", \"T\") == \"T\", \"Expected: 'A' or 'T', got: \" + str(merge('A', 'T'))\n",
    "assert merge(\"-A\", \"T-\") == \"TA\", \"Expected: 'TA', got: \" + str(merge('-A', 'T-'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You have accomplished a lot today! Well done! You've implemented all of the building blocks you will require to complete the project. Tomorrow you will be able to generate sequence assemblies using what you created today, brought into existance through sheer power of your will!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Part 3: Bringing everything together\n",
    "\n",
    "You have laid all the ground work. Done all of the hard work. Now all you need to do is stitch everything you've done together and produce a sequence assembly. Let's review the steps in the algorithm...\n",
    "\n",
    "1. Сalculate pairwise alignments of all fragments.\n",
    "1. Choose two fragments with the largest overlap.\n",
    "1. Merge chosen fragments.\n",
    "1. Repeat step 2 and 3 until only one fragment is left or you cannot merge anymore.\n",
    "\n",
    "You can translate this into a pseudocode like this:\n",
    "\n",
    "1. Loop number of fragments times\n",
    "    1. Calculate all pairwise alignments between fragments with `align()`\n",
    "    1. Pick the _best_ alignment using `score()`\n",
    "    1. Merge the alignment into a contig using `merge()`\n",
    "    1. Remove the fragments that generated the alignment and put the contig back into the collection of fragments\n",
    "    \n",
    "The supplied `project` module defines a slightly more sophisticated alignment generating algorithm that will produce better results. You're welcome to use this if you wish, it is used in the template function but if you wish to use your own alignment generation function simply remove the import."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from solutions.sa import align, needleman_wunsch, score\n",
    "\n",
    "def assemble(frags):\n",
    "    n = len(frags)\n",
    "\n",
    "    for _ in range(n - 1):\n",
    "        # Generate all possible alignments\n",
    "        # Generate each alignment with align(frag1, frag2, needleman_wunsch)\n",
    "        \n",
    "        # Pick the \"best\" alignment\n",
    "        \n",
    "        # Merge the best alignment into a contig\n",
    "        contig = merge(_, _)\n",
    "        \n",
    "        # Remove fragments you just merged together\n",
    "        \n",
    "        # Add contig you just created with merge()\n",
    "        \n",
    "    return frags[0]\n",
    "\n",
    "sequence = assemble(fragments.copy())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If everything has worked correctly you will now have a nice assembled sequence!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "One you are satisfied with your assembled sequence, paste it into the [BLAST](https://blast.ncbi.nlm.nih.gov/Blast.cgi) search form to see if you can identify the gene.\n",
    "\n",
    "For extra points, ensure you add docstrings for each function. You may also with to experiment with using `edit_distance()` instead of `needleman_wunsch()`. What difference, if any, does it make?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
